### PR TITLE
fix(registrar): protect cached discovery signers

### PR DESCRIPTION
**Pull request by leopoldjoy**
## Summary
- Add a run-local last-known-active signer cache for registrar discovery cycles.
- Preserve cached signers and mark their instance unresolved when an instance disappears from otherwise successful discovery output.
- Evict cached missing instances after `INSTANCE_CACHE_TTL_CYCLES` so real removals eventually cancel proofs and deregister signers.
- Add driver regression tests for missing-instance protection, TTL eviction, TTL reset on refresh, and cold-start empty discovery behavior.

## Trade-off
- The 5-cycle TTL reduces transient discovery flakes causing proof cancellation or orphan deregistration.
- If the TTL is too short, partial discovery glitches can still leak through.
- If the TTL is too long, genuine instance removal cleanup is delayed.

## Testing
- `cargo +nightly fmt --all`
- `RISC0_SKIP_BUILD_KERNELS=1 cargo clippy -p base-proof-tee-registrar --all-targets -- -D warnings`
- `RISC0_SKIP_BUILD_KERNELS=1 cargo test -p base-proof-tee-registrar`
- Local caveat: exact cargo commands without `RISC0_SKIP_BUILD_KERNELS=1` fail on this machine because `risc0-sys` invokes `xcrun metal`, and the `metal` developer tool is not installed or not on PATH.

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>✅ Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
✅
<code>1/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td><code>2</code> if repo is sensitive</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: crates/proof/tee/registrar/src/driver.rs [hunk 1]
```diff
@@ -41,6 +45,14 @@ pub const DEFAULT_MAX_CONCURRENCY: usize = 4;
 /// of 90 minutes.
 pub const DEFAULT_UNHEALTHY_REGISTRATION_WINDOW_SECS: u64 = 5100;
 
+/// Number of consecutive discovery cycles to protect last-known active signers
+/// for an instance that disappears from otherwise successful discovery output.
+///
+/// Five cycles is roughly 2.5 minutes with the default 30 second poll interval.
+/// A shorter window is more vulnerable to transient discovery flakes; a longer
+/// window delays cleanup when an instance was genuinely removed.
+pub const INSTANCE_CACHE_TTL_CYCLES: u32 = 5;
```
[the remark this answers is not in the document]
Comment (leopoldjoy) [reply crates/proof/tee/registrar/src/driver.rs:54]: Keeping TTL as a documented policy constant for this PR to avoid widening config surface; happy to make it configurable separately if operators need tuning.

File: crates/proof/tee/registrar/src/driver.rs [hunk 2]
```diff
@@ -335,6 +353,12 @@ where
...
                 Ok(outcome) => {
+                    let active_signers = outcome.active_signers.iter().copied().collect::<Vec<_>>();
+                    if active_signers.is_empty() {
+                        last_known_active.remove(&instance.instance_id);
+                    } else {
+                        last_known_active.insert(instance.instance_id, (active_signers, 0));
+                    }
                     resolution.registerable.extend(outcome.registerable);
                     resolution.active_signers.extend(outcome.active_signers);
                     resolution.unresolved_instance_ids.extend(outcome.unresolved_instance_ids);
```
[the remark this answers is not in the document]
Comment (leopoldjoy) [reply crates/proof/tee/registrar/src/driver.rs:368]: Addressed in df54d30 by resetting cached TTL on resolve errors and adding a regression test documenting that TTL only ages while discovery omits the instance.

File: crates/proof/tee/registrar/src/driver.rs [hunk 3]
```diff
@@ -335,11 +353,23 @@ where
...
                 }
                 Err(e) => {
+                    // Resolution failures are handled by `unresolved_instance_ids` for this
+                    // cycle; the last-known cache only ages while discovery omits the instance.
+                    let cache_entry = last_known_active.get_mut(&instance.instance_id);
```
[the remark this answers is not in the document]
Comment (leopoldjoy) [reply crates/proof/tee/registrar/src/driver.rs]: Addressed in a0e73b9 by extending cached signers into `active_signers` on resolve errors and asserting that behavior in the existing regression test.

File: crates/proof/tee/registrar/src/driver.rs [hunk 4]
```diff
@@ -339,11 +353,24 @@ where
...
+                        last_known_active.remove(&instance.instance_id);
+                    } else {
+                        last_known_active.insert(instance.instance_id, (active_signers, 0));
+                    }
                     resolution.registerable.extend(outcome.registerable);
                     resolution.active_signers.extend(outcome.active_signers);
                     resolution.unresolved_instance_ids.extend(outcome.unresolved_instance_ids);
                 }
                 Err(e) => {
+                    // Resolution failures are handled by `unresolved_instance_ids` for this
```
Comment (jackchuma) [inline crates/proof/tee/registrar/src/driver.rs]: delete: the cached-signer path for resolve errors is extra state handling that does not change the important outcomes. A resolve error already adds the instance to `unresolved_instance_ids`; that already prevents orphan deregistration for the cycle and preserves pending proof tasks via `SignerManager::reconcile_proof_tasks`. Extending `active_signers` from the cache is therefore redundant for current consumers

File: crates/proof/tee/registrar/src/driver.rs [hunk 5]
```diff
@@ -356,6 +383,36 @@ where
             }
         }
 
+        let mut expired_instance_ids = Vec::new();
```
Comment (jackchuma) [inline crates/proof/tee/registrar/src/driver.rs]: stdlib: the `expired_instance_ids` Vec exists only because the map cannot be removed from while iterating mutably. `HashMap::retain` is the standard tool for this: age missing entries, preserve active ones, log expired ones, and return `false` for expired entries. Same behavior, no temporary Vec and no second loop

File: crates/proof/tee/registrar/src/driver.rs [hunk 6]
```diff
@@ -598,4 +661,135 @@ mod tests {
             );
         }
     }
+
+    #[tokio::test]
+    async fn discover_and_resolve_protects_cached_instance_missing_from_discovery() {
```
Comment (jackchuma) [inline crates/proof/tee/registrar/src/driver.rs]: delete: the one-cycle cache test is covered by the TTL test immediately below. That TTL test’s first loop iteration already verifies TTL `1`, cached signer preservation, and unresolved marking. If the `registerable.is_empty()` assertion matters, add that one assertion to the TTL test instead of keeping a whole separate test

File: crates/proof/tee/registrar/src/driver.rs [hunk 7]
```diff
@@ -598,4 +661,135 @@ mod tests {
...
+
+        let expired_resolution =
+            missing_cycle.discover_and_resolve(&mut last_known_active).await.unwrap();
+
+        assert!(expired_resolution.active_signers.is_empty());
+        assert!(expired_resolution.unresolved_instance_ids.is_empty());
+    }
+
+    #[tokio::test]
+    async fn discover_and_resolve_resolve_error_resets_cached_instance_ttl() {
```
Comment (jackchuma) [inline crates/proof/tee/registrar/src/driver.rs]: delete: this test only exists to pin the redundant resolve-error cache reset from the first finding. The useful behavior is already covered elsewhere: unresolved instances preserve proof tasks in `signer_manager.rs`, and the driver already has a test that unreachable instances become unresolved. Remove the special TTL-reset behavior and this test goes with it

File: crates/proof/tee/registrar/src/driver.rs (hunk 5 above)
Comment (leopoldjoy) [reply crates/proof/tee/registrar/src/driver.rs]: Done in 74d4fd6: replaced the temporary expired-ID vec and second removal loop with `HashMap::retain`.

File: crates/proof/tee/registrar/src/driver.rs (hunk 4 above)
Comment (leopoldjoy) [reply crates/proof/tee/registrar/src/driver.rs]: Done in 74d4fd6: removed the resolve-error cache special handling so existing `unresolved_instance_ids` behavior remains the only protection for that path.

File: crates/proof/tee/registrar/src/driver.rs (hunk 6 above)
Comment (leopoldjoy) [reply crates/proof/tee/registrar/src/driver.rs]: Done in 74d4fd6: removed the duplicate one-cycle test and folded the `registerable.is_empty()` assertion into the TTL test.

File: crates/proof/tee/registrar/src/driver.rs (hunk 7 above)
Comment (leopoldjoy) [reply crates/proof/tee/registrar/src/driver.rs]: Done in 74d4fd6: removed the resolve-error TTL test along with the redundant behavior it pinned.

File: crates/proof/tee/registrar/src/driver.rs [hunk 8]
```diff
@@ -595,4 +651,79 @@ mod tests {
...
+            let resolution =
+                missing_cycle.discover_and_resolve(&mut last_known_active).await.unwrap();
+
+            assert!(resolution.registerable.is_empty());
+            assert!(resolution.active_signers.contains(&signer_addr));
+            assert_eq!(
+                resolution.unresolved_instance_ids,
+                HashSet::from([inst.instance_id.clone()])
+            );
+            assert_eq!(
```
[the remark this answers is not in the document]
Comment (leopoldjoy) [reply crates/proof/tee/registrar/src/driver.rs:695]: Actionable and fixed in 504b631e: updated the TTL eviction test to use `MockEnclaveEndpointClient` after merging current `main`.

File: crates/proof/tee/registrar/src/driver.rs [hunk 9]
```diff
@@ -595,4 +651,79 @@ mod tests {
...
+
+        let refresh_resolution =
+            present_cycle.discover_and_resolve(&mut last_known_active).await.unwrap();
+
+        assert!(refresh_resolution.active_signers.contains(&signer_addr));
+        assert!(refresh_resolution.unresolved_instance_ids.is_empty());
+        assert_eq!(last_known_active.get(&inst.instance_id).map(|(_, ttl)| *ttl), Some(0));
+
+        for expected_ttl in 1..=INSTANCE_CACHE_TTL_CYCLES {
+            let resolution =
```
[the remark this answers is not in the document]
Comment (leopoldjoy) [reply crates/proof/tee/registrar/src/driver.rs:731]: Actionable and fixed in 504b631e: updated the TTL refresh test to use `MockEnclaveEndpointClient` after merging current `main`.

**Review by jackchuma — APPROVED** [review]
